kedro/io/catalog_config_resolver.py

```diff
@@ -0,0 +1,261 @@
+            raise ValueError(
+                f"Multiple catch-all patterns found in the catalog: {', '.join(catch_all)}. Only one catch-all pattern is allowed, remove the extras."
+            )
+        return {key: dataset_patterns[key] for key in sorted_keys}
+
+    def list_patterns(self) -> list[str]:
+        """List al patterns available in the catalog."""
+        return (
+            list(self._dataset_patterns.keys())
+            + list(self._default_pattern.keys())
+            + list(self._runtime_patterns.keys())
+        )
+
+    def match_pattern(self, ds_name: str) -> str | None:
+        """Match a dataset name against patterns in a dictionary."""
+        all_patterns = self.list_patterns()
+        matches = (pattern for pattern in all_patterns if parse(pattern, ds_name))
+        return next(matches, None)
+
+    @classmethod
+    def _extract_patterns(
+        cls,
+        config: dict[str, dict[str, Any]] | None,
+        credentials: dict[str, dict[str, Any]] | None,
+    ) -> tuple[Patterns, Patterns]:
+        """Extract and sort patterns from the configuration."""
+        config = copy.deepcopy(config) or {}
+        credentials = copy.deepcopy(credentials) or {}
+        dataset_patterns = {}
+        user_default = {}
+
+        for ds_name, ds_config in config.items():
+            if cls.is_pattern(ds_name):
+                dataset_patterns[ds_name] = _resolve_credentials(ds_config, credentials)
+
+        sorted_patterns = cls._sort_patterns(dataset_patterns)
+        if sorted_patterns:
+            # If the last pattern is a catch-all pattern, pop it and set it as the default
+            if cls._pattern_specificity(list(sorted_patterns.keys())[-1]) == 0:
+                last_pattern = sorted_patterns.popitem()
+                user_default = {last_pattern[0]: last_pattern[1]}
+
+        return sorted_patterns, user_default
+
+    def _init_configs(
+        self,
+        config: dict[str, dict[str, Any]] | None,
+        credentials: dict[str, dict[str, Any]] | None,
+    ) -> dict[str, dict[str, Any]]:
+        """Initialize the dataset configuration with resolved credentials."""
+        # TODO: check if deep copies are required
+        config = copy.deepcopy(config) or {}
+        credentials = copy.deepcopy(credentials) or {}
+        resolved_configs = {}
+
+        for ds_name, ds_config in config.items():
+            if not isinstance(ds_config, dict):
+                raise DatasetError(
+                    f"Catalog entry '{ds_name}' is not a valid dataset configuration. "
+                    "\nHint: If this catalog entry is intended for variable interpolation, "
+                    "make sure that the key is preceded by an underscore."
+                )
+            if not self.is_pattern(ds_name):
+                resolved_configs[ds_name] = _resolve_credentials(ds_config, credentials)
+
+        return resolved_configs
+
+    def resolve_dataset_pattern(
```

Comment:
I know this code is not yours and simply moved here, but I can see that we should try to simplify it a bit (maybe in a future PR). I might be wrong and underestimating the problem, but it looks unnecessary convoluted to me.

Reply:
Slightly simplified it